autogen/agentchat/contrib/society_of_mind_agent.py

```diff
@@ -0,0 +1,116 @@
+        chat_manager (GroupChatManager): the group chat manager that will be running the inner monologue
+        response_preparer (Optional, Callable): A function reference of the form:
+                    f( self: SocietyOfMindAgent, messages: List[Dict])
+                Where self is this SocietyOfMindAgent, and messages is a list of inner-monologue messages. The
+                function should return a string representing the final response (extracted or prepared) from
+                that history. The default response_preparer deterministically returns the last message in the
+                inner-monolgue, but strips out any termination signals.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        chat_manager: GroupChatManager,
+        response_preparer: Optional[Union[str, Callable]] = lambda self, messages: messages[-1]["content"]
+        .replace("TERMINATE", "")
+        .strip(),
+        is_termination_msg: Optional[Callable[[Dict], bool]] = None,
+        max_consecutive_auto_reply: Optional[int] = None,
+        human_input_mode: Optional[str] = "TERMINATE",
+        function_map: Optional[Dict[str, Callable]] = None,
+        code_execution_config: Optional[Union[Dict, Literal[False]]] = None,
+        llm_config: Optional[Union[Dict, Literal[False]]] = False,
+        default_auto_reply: Optional[Union[str, Dict, None]] = "",
+    ):
+        super().__init__(
+            name=name,
+            system_message="",
+            is_termination_msg=is_termination_msg,
+            max_consecutive_auto_reply=max_consecutive_auto_reply,
+            human_input_mode=human_input_mode,
+            function_map=function_map,
+            code_execution_config=code_execution_config,
+            llm_config=llm_config,
+            default_auto_reply=default_auto_reply,
+        )
+
+        self.update_chat_manager(chat_manager)
+        self.response_preparer = response_preparer
+
+        self._reply_func_list = []
+        self.register_reply([Agent, None], SocietyOfMindAgent.generate_inner_monologue_reply)
+        self.register_reply([Agent, None], ConversableAgent.generate_code_execution_reply)
+        self.register_reply([Agent, None], ConversableAgent.generate_function_call_reply)
+        self.register_reply([Agent, None], ConversableAgent.check_termination_and_human_reply)
+
+    @property
+    def chat_manager(self) -> Union[GroupChatManager, None]:
+        """Return the group chat manager."""
+        return self._chat_manager
+
+    def update_chat_manager(self, chat_manager: Union[GroupChatManager, None]):
+        """Update the chat manager.
+
+        Args:
+            chat_manager (GroupChatManager): the group chat manager
+        """
+        self._chat_manager = chat_manager
+
+        # Awkward, but due to object cloning, there's no better way to do this
+        # Read the GroupChat object from the callback
+        self._group_chat = None
```

Comment:
Should we just change `GroupChatManager` instead to have `self._group_chat = group_chat` inside the init function?

Reply:
I don't think we can. It's related to this: https://github.com/microsoft/autogen/pull/961#discussion_r1425813041

If GroupChat is accessed from anywhere but the from the *copy* in the registered reply config, then there is a chance it diverges from what GroupChatManager is actually using. Also related is #1006 and #409 

There are a lot of very nuanced and profound implications for how that composability mechanism is implemented, and how it is exposed via GroupChatManager. I've shot myself in the foot about a half dozen times because of this.